### PR TITLE
Fix T6882 (async functions are not hoisted)

### PR DESCRIPTION
**Pull request by erikdesjardins**
https://phabricator.babeljs.io/T6882

Reverting #3097 fixes T6882, but breaks [T3085](https://phabricator.babeljs.io/T3085) and [T3026](https://phabricator.babeljs.io/T3026) (duplicates).

I don't think that the way I fixed T3085/T3026 is ideal, but I feel that it's preferable to turning off hoisting for _all_ async function declarations (à la #3097).

### CODE REVIEW

**Comment by hzoo** [comment]
Related was https://github.com/babel/babel/pull/3118, we'll have to look at all this again

File: packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/statement/actual.js [hunk 1]
```diff
@@ -1,5 +1,3 @@
-function normalFunction() {}
```
Comment (kittens) [inline packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/statement/actual.js]: Why was this removed?

Comment (hzoo) [inline packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/statement/actual.js]: Ah it was since he did revert as the first commit - https://github.com/erikdesjardins/babel/commit/ec61bd9386c86fc4006e121b9ad6bdc9b7b2c8f0

Comment (erikdesjardins) [inline packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/statement/actual.js]: Yep - it was there to test that the async function _wasn't_ hoisted, which is incorrect.

I could put it back (and move it below the async function in the output), but that should be covered by e759b5f835235ea65948f9e35073ab76dec8b5c3.

**Comment by hzoo** [comment]
I guess I'm ok with this?

**Comment by hzoo** [comment]
@amasad good?

**Comment by amasad** [comment]
LGTM

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js [hunk 2]
```diff
@@ -399,14 +399,21 @@ export default function () {
               hoistedExportsNode = buildExportsAssignment(t.identifier(name), hoistedExportsNode).expression;
             }
 
-            topNodes.unshift(t.expressionStatement(hoistedExportsNode));
+            const node = t.expressionStatement(hoistedExportsNode);
+            node._blockHoist = 3;
```
Comment (hzoo) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: probably don't need a comment (assuming you know what `_blockHoist` does)

**Comment by hzoo** [comment]
Thanks @erikdesjardins!

**Comment by erikdesjardins** [comment]
Thank *you* :)

File: packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js (hunk 2 above)
Comment (erikdesjardins) [inline packages/babel-plugin-transform-es2015-modules-commonjs/src/index.js]: Ah, yeah. hopefully the blame will be sufficient.
